ygot/struct_validation_map.go

```diff
@@ -319,3 +321,189 @@ func MergeJSON(a, b map[string]interface{}) (map[string]interface{}, error) {
+		if err := copyStruct(d.Elem(), srcField.Elem()); err != nil {
+			return err
+		}
+		dstField.Set(d)
+	default:
+		p := reflect.New(srcField.Type().Elem())
+		p.Elem().Set(srcField.Elem())
+		dstField.Set(p)
+	}
+
+	return nil
+}
+
+// copyInterfaceField copies srcField into dstField. Both srcField and dstField
+// are reflect.Value structs which contain an interface value.
+func copyInterfaceField(dstField, srcField reflect.Value) error {
+	if srcField.IsNil() || !srcField.IsValid() {
+		return nil
+	}
+
+	if srcField.Elem().Kind() != reflect.Ptr {
+		return fmt.Errorf("invalid interface type received: %T", srcField.Interface())
+	}
+
+	switch srcField.Elem().Elem().Kind() {
+	case reflect.Struct:
+		var d reflect.Value
+		d = reflect.New(srcField.Elem().Elem().Type())
+		if err := copyStruct(d.Elem(), srcField.Elem().Elem()); err != nil {
+			return err
+		}
+		dstField.Set(d)
+	default:
+		return fmt.Errorf("unknown type in interface, %s", srcField.Elem().Kind())
+	}
+	return nil
+}
+
+// copyMapField copies srcField into dstField. Both srcField and dstField are
+// reflect.Value structs which contain a map value. If dstField is populated
+// then an error is returned.
+// TODO(robjs): Implement merging of maps when they are populated in the dstField
+// supplied. See https://github.com/openconfig/ygot/issues/74.
+func copyMapField(dstField, srcField reflect.Value) error {
+	if srcField.Len() == 0 {
+		return nil
+	}
+
+	if dstField.Len() != 0 {
+		return fmt.Errorf("unimplemented: cannot map slice where destination was set: %v == %v", srcField.Type().Name(), dstField.Interface())
+	}
+
+	keys := srcField.MapKeys()
+	if k := srcField.MapIndex(keys[0]).Kind(); k != reflect.Ptr {
+		return fmt.Errorf("invalid map, got member type %s", k)
+	}
+
+	// TODO(robjs): When we move to go1.9+ only support in ygot, we can use the
+	// following to make a map of a particular size.
+	// nm := reflect.MakeMapWithSize(reflect.MapOf((keys[0]).Type(), srcField.MapIndex(keys[0]).Type()), srcField.Len())
+	nm := reflect.MakeMap(reflect.MapOf((keys[0]).Type(), srcField.MapIndex(keys[0]).Type()))
+	for _, k := range keys {
+		v := srcField.MapIndex(k)
+		d := reflect.New(v.Elem().Type())
+		if err := copyStruct(d.Elem(), v.Elem()); err != nil {
+			return err
+		}
+		nm.SetMapIndex(k, d)
+	}
+	dstField.Set(nm)
+	return nil
+}
+
+// copySliceField copies srcField into dstField. Both srcField and dstField are
+// reflect.Value structs which represent slices. If the slice in dstField is populated
+// an error is returned.
+// TODO(robjs): Implement merging of slice fields when they are populated in the
+// dstField, see https://github.com/openconfig/ygot/issues/74.
+func copySliceField(dstField, srcField reflect.Value) error {
+	if dstField.Len() != 0 {
+		return fmt.Errorf("unimplemented: cannot map slice where destination was set: %v == %v", srcField.Type().Name(), dstField.Interface())
+	}
+
+	switch {
+	case srcField.Type().Elem().Kind() != reflect.Ptr:
```

Comment:
Same suggestions re: Elem() as above.

Reply:
I don't think we can do that here, because we need to check that this thing actually isn't a slice of struct pointers, adopted a method from `util/reflect.go` that makes this clearer.